design-proposals/2023-02-08-code-organisation.md

```diff
@@ -0,0 +1,137 @@
+
+- widgets
+- gestures support on toucscreen devices
+- location services (GPS, location from the mobile network)
+
+### Organisation
+
+MapLibre core could be provided as a submodule, but it is not necessary. For mobile platforms this may or may not be the case, depending on the level of the integration, for example:
+
+- on Android if Kotlin/Java is mostly used, then distributing the core separately as a dynamic library is completely reasonable
+- on iOS where static building may be much more beneficial due to size the core will probably be built together with the SDK
+
+We leave the SDK organisation up to the specific platform developers. It should probably follow conventions of the platform to simplify distribution and usage.
+
+The build system is also not explicitly defined. It would be strongly preferred that the core is built by `CMake` (or at least the necessary build scripts are not duplicated in a SDK repository/folder but generated by `CMake`).
+
+### Scope and feature-parity
+
+Different SDKs and bindings **are not expected to have feature-parity**. They can/will live in separate repositories and will be developed independently.
+
+When a feature is added to the core it may not be added to all SDKs at the same time.
+
+## Repository split - yes or no?
+
+This proposal does not touch much on repository splitting. If SDKs continue to live in the main repository, they should be put under `sdk/<platform>`. The following primary SDKs should be created and supported based on the currently available code:
+
+- Android SDK
+- Apple SDK (common between all Apple platforms)
+- Qt SDK
+- Node.js Binaries and SDK
+- GLFW SDK
+
+Navigation SDKs could then be merged with respective platform SDK.
+
+### Repository naming
+
+- `maplibre-native-core` or `maplibre-native-renderer` for the core functionality.
+- SDKs:
+  - `maplibre-native-sdk-apple` (iOS and macOS, potentially merged with `maplibre-navigation-ios`)
+  - `maplibre-native-sdk-android` (Android, potentially merged with `maplibre-navigation-android`)
+  - `maplibre-native-sdk-qt` (Qt bindings, should probably be merged with `qt-geoservices-maplibre-gl`)
+  - `maplibre-native-sdk-js`
+
+Alternatively SDKs could be without the explicit `sdk` in the name.
+
+## How to organise the work?
+
+This has to be discussed a bit more but the following rough proposal could be done in case of repository split:
+
+1. A stable release is tagged for all platforms that we currently support.
+
+2. A new repository `maplibre-native-core` is created and the core cleanup is done there. `maplibre-gl-native` is kept as a snapshot or as a `stable` branch as part of the `maplibre-native-core`.
+
+3. The proposal is implemented for the core.
+
+4. SDK repositories are created on by one as needed and they migrate to use `maplibre-native-core`.
```

Comment:
Will any new repository keep the commit history or start over from scratch? There are a number of tools such as [git-filter-repo](https://github.com/newren/git-filter-repo/) for creating a new repository that contains just the history relevant to a given path.

Reply:
Excellent idea!

Reply:
Splitting out a separate repository for platform-specific code has several benefits that only apply if commit history outside the platform/*/ directory gets filtered out:

* It’ll be a lot easier to find platform-specific changes without noise from other platforms or core.
* Contributors to core would be less likely to make unnecessary changes to platform code, such as in #736.[^gcc]
* Contributors who primarily work on platform-specific code can get away with only making a shallow clone of the main gl-native repository, avoiding a lot of redundant disk space usage.
* The maplibre-gl-native-distribution repository can be folded back into a maplibre-gl-native-ios repository, simplifying the release process, improving discovery by downstream developers, and simplifying the installation steps.
* Source-based distribution would be infeasible in a monorepo or even a separate repository containing the monorepo’s history, because downstream developers would have to endure a ton of unnecessary cloning. This would be especially problematic with SPM, which tends to reclone the repository quite often and disables Xcode functionality until it’s done cloning.
* One platform’s SDK could more easily pull in dependencies using the platform’s usual package managers instead of relying on the lowest common denominator of package managers, Git submodules.

Unfortunately, it looks like history has already been broken more than once for a large number of platform-specific files. Git can only track history across file moves, but not manual copy-pasting across separate working copies. This breakage is only a minor annoyance if the repository remains a monorepo forever, but I think it would be prudent not to make that assumption. If platform-specific code is ever split back out into a separate repository at some point in the future, the filtering I’m proposing here would actually cause history to be lost.

[^gcc]: No one uses GCC to compile AppKit and UIKit code these days. It’s completely unnecessary to make the Objective-C coding style conform to GCC’s C/C++ warnings. This case of repository-wide find-and-replace was harmless, but there may be other gotchas in the future.

Reply:
Would you be open to the idea of starting over, so to speak?

The platform/ios/ directory was copied over from a separate mapbox-gl-native-ios repository, presumably at the last commit before the license changed. (It would be good to confirm that.) Not too many commits have touched this directory since, only a hundred or so, and many were only necessary in order to make the code build in a monorepo again.

Consider a workflow along these lines:

1. Create a new maplibre-gl-native-ios repository (or whatever you’d like to call it). Move aside the main branch.
2. Add mapbox-gl-native-ios as a remote.
3. Check out the last commit of mapbox-gl-native-ios that was permissively licensed, as well as any ios-* and macos-* release tags up to that point.
4. Create a new main branch from that commit.
5. [Filter the repository](https://github.com/newren/git-filter-repo/) to only include platform/{darwin,ios,macos}/ and certain files in the root directory, like .gitmodules, Cartfile, and LICENSE.md.
6. Add maplibre-gl-native as a remote.
7. Cherry-pick relevant [commits from platform/ios/ in maplibre-gl-native](https://github.com/maplibre/maplibre-gl-native/commits/main/platform/ios/) into the new branch. Git should be able to automatically resolve discrepancies like the redundant platform/ios/ in maplibre-gl-native.
   * If cherry-picking would be too time-consuming, rebasing would work too, but you’ll have to undo any monorepo-specific changes that would now be counterproductive.
8. Push the tags and copy over any ios-* and macos-* releases.
9. Delete platform/ios/ and any ios-* and macos-* tags and releases from maplibre-gl-native.

You could clean up the submodules afterwards. It’d be a lot easier than cleaning up the history of .gitmodules in a massive interactive rebase.

Reply:
I am very much aligned with @1ec5 here.

Reply:
Picking up on this. Should this be added to the proposal or is it out of scope here?